Use installation prefix inside C sources

mwperina opened the pull request:
Propagate actual directories depending on installation prefix into C
sources.
For example:

* RPM installations - sysconfdir contains '/etc'
* meson installtions - sysconfdir contains '/usr/local/etc'

Signed-off-by: Martin Perina <mperina@redhat.com>

alexlarsson commented on src/libhirte/common/paths.h.in:
```diff
@@ -0,0 +1,9 @@
+
+/* Standard configuration file locations
+ */
+#define CFG_ETC_HIRTE_CONF "@SYSCONFDIR@/hirte/hirte.conf"
```
This is a bit weird way of doing this. The typical way is to make SYSCONFDIR a c macro from meson, which will eventually be passed via a -DSYSCONF=\"foo\" compiler option, and then you use it from the C code using string merging like this:

#define CFG_ETC_HIRTE_CONF SYSCONFDIR "/hirte/hirte.conf"

mwperina replied:
Yeah, I've already found that out, but fighting to aling meson source generation with lint working properly. But it seems that a few minutes ago I found out the working solution ...

mwperina committed:
Propagate build time configuration to C sources

1. Autogenerate config.h header which will include directory definitions
   based on the prefix used in the build

2. Add dependency of existing configuration files paths to the prefix
   used during the build (the prefix is different between RPM build and
   regular meson build).

Signed-off-by: Martin Perina <mperina@redhat.com>

mwperina committed:
Fix incorrect order of headers in service.c

Signed-off-by: Martin Perina <mperina@redhat.com>

mwperina committed:
Add build dependencies to lint GH job

Because we generate header files as a part of the build, build
dependencies are required on the lint job

Signed-off-by: Martin Perina <mperina@redhat.com>

alexlarsson commented on meson.build:
```diff
@@ -29,6 +29,26 @@ if api_bus == 'user'
+sysconfdir = get_option('sysconfdir')
+if not sysconfdir.startswith('/')
+    # sysconfdir is relative to the prefix
+    sysconfdir = prefixdir / get_option('sysconfdir')
```
This is a bit overcomplex, this should be enough:
```
sysconfdir = join_paths(prefix, get_option('sysconfdir'))
```

alexlarsson replied:
As in, join_paths returns the second arg if it is an absolute path already.

mwperina replied:
Fixed

alexlarsson commented on meson.build:
```diff
@@ -29,6 +29,26 @@ if api_bus == 'user'
+        configuration : path_conf)
+
+# For lint to work properly the same include needs to be added manually to clang-tidy arguments
+add_project_arguments('-include', 'path.h', language : 'c')
```
'config.h' is the more traditional name, in case we need to set other optional configuration defines.

mwperina replied:
I know, but we already have `libhirte/common/config.h`, so wouldn't it be confusing to have 2 different `config.h` files in the project?

alexlarsson replied:
Maybe we can call that something else? cfg.h would match the prefix used for the function.

engelmi commented on meson.build:
```diff
@@ -29,6 +29,29 @@ if api_bus == 'user'
+path_conf.set_quoted('SYSCONF_DIR', sysconfdir)
+
+template = 'sysconfdir = "@0@"'
+message(template.format(sysconfdir))
```
This is just for debugging and could be removed.

mwperina replied:
Yes, I'm going to remove that

engelmi reviewed (commented):
Is it really necessary to generate a dedicated file for the `sysconfdir`? Couldn't we just use a [global argument](https://mesonbuild.com/Adding-arguments.html) like we already do it for the GNU_SOURCE:
https://github.com/containers/hirte/blob/main/meson.build#L16-L17
For example, have 
```
add_global_arguments('-DSYSCONF_DIR', language : 'c')
```

mwperina committed:
Move _GNU_SOURCE macro to generated config.h

Moves _GNU_SOURCE macro to generated config.h to align with other
existing build times macros in config.h

Signed-off-by: Martin Perina <mperina@redhat.com>

engelmi reviewed (approved):
LGTM
